### PR TITLE
Add appProtocol to EndpointSlice.Port

### PR DESCRIPTION
**Pull request by howardjohn**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
This optional field will allow specifying information about the
application protocol used by the endpoint, such as http, grpc, etc. This
field is a string such arbitrary protocols can be specified.

**Does this PR introduce a user-facing change?**:

```release-note
Added appProtocol field to EndpointSlice Port
```

### CODE REVIEW

**Comment by robscott** [comment]
Hey @howardjohn, this looks like a great start, thanks! One thing I forgot to mention is code generation. For e2e tests to pass and things to build, you'll need to run `make update` first and commit/amend the changes. Usually `hack/update-bazel.sh` is sufficient for small changes, but when types are added/changed, I usually end up having to run a full `make update`.

**Comment by howardjohn** [comment]
@robscott thanks, I had some issues running it but I think I got everything to run

**Comment by robscott** [comment]
Thanks for the work on this! This is in the process of getting formal approval in an associated KEP PR here: https://github.com/kubernetes/enhancements/pull/1286. I think everyone's on board with this, but I don't think we'll be able to merge this until the associated enhancement PR gets merged.

File: pkg/apis/discovery/types.go
Comment (freehan) [inline pkg/apis/discovery/types.go]: Need to have more detail comment about the format and default values. Plus examples

Comment (robscott) [reply pkg/apis/discovery/types.go]: @freehan @howardjohn what formats should we allow here? I think the label validation that John added is a good start here. Should we limit this to a length shorter than the current 63 characters we get from that validation? Would a protocol ever include non alphanumeric characters? Should we enforce all uppercase or lowercase values so we don't get different `https` and `HTTPS` values? I think we agreed that the default value is nil/there isn't one, right?

My naive thought would be to leave the upper limit at 63, require alphanumeric characters only, and require uppercase letters, but I'm very open to other ideas.

Comment (robscott) [reply pkg/apis/discovery/types.go]: @thockin Any thoughts on how we should limit/validate AppProtocol?

Comment (howardjohn) [reply pkg/apis/discovery/types.go]: My thoughts:
* Length limit - 63 seems reasonable, I cannot imagine needing a longer protocol really and its pretty standard across Kubernetes
* Character set: I assume values like `mycompany.com/our-custom-protocol` are not something that would really be used here. If that is the case, all uppercase sounds good. We may also want to support a `-`?

Some examples of things someone might want to set with various constraints:
`DNSOVERHTTPS`, `DNS-OVER-HTTPS`, `DNS-Over-HTTPS`, `dns-over-https`
`HTTP`, `http`
`MYSQL`, `MY-SQL`, `mySql`, `mysql`

Based on this, I think `-` is required, but I don't think there is value to mixing cases.

And agree, default value is `nil`

Comment (thockin) [reply pkg/apis/discovery/types.go]: I think there are 2 options:

1) Treat these as constants.  Constants start with capital letters, acronyms are in caps.  e.g. `HTTP`, `DNSOverHTTPS`, `MySQL`, `MyCustomProtocol`

2) Treat this as a label key.  Un-prefixed names would be reserved for standardized values, prefixed names would be for non-standard values.  E.g. `http`, `dns-over-https`, `mysql`, `mycompany.com/my-custom-protocol`

I don't think any other options make sense.  Between those two, I don't feel VERY strongly - I have a slight lean toward label-style, but it's a little weird if we expect the vast majority of use-cases to be the non-prefixed.  Which do we expect more of?

Comment (robscott) [reply pkg/apis/discovery/types.go]: Very interesting. How would we validate that second approach? Would we have a whitelist for anything without a domain prefix and then anything with a prefix like that is valid? Or do we just provide guidelines such as 1 or 2 and only enforce something as simple as max/min length and characters allowed? Any kind of whitelist here would be challenging to change post-alpha assuming it would be similarly [complex as adding addressTypes](https://github.com/kubernetes/kubernetes/pull/84091#discussion_r338268424).

Comment (howardjohn) [reply pkg/apis/discovery/types.go]: I am not sure that standardized values is the best approach here. There are a lot of protocols out there (wikipedia [lists about 50](https://en.wikipedia.org/wiki/Application_layer) and there are of course many more less standardized one). I think it would be hard to hit the sweet spot of having the list of "standard" protocols not too small or too large. It would also be a bit awkward UX to be able to specify `MySQL` but then I have another database and now I have to specify `postgresql.org/psql` or something. It seems a bit arbitrary. What do you think?

Since the protocol is given meaning by the application consuming it, I think just allowing arbitrary strings here makes the most sense.

I don't have any opinion on the exact character set here though, as long as its not just upper case letters (`DNSOVERHTTPS` is awkward)

Comment (robscott) [reply pkg/apis/discovery/types.go]: I agree with that. It would likely not make sense for us to try to come up with a set of "standard" app protocols here, and instead it likely makes sense to have actual validation be limited to a simple character set and max length. In addition to that, it likely makes sense to provide guidance on how protocols should be represented, maybe with some specific examples for common app protocols?

Comment (robscott) [reply pkg/apis/discovery/types.go]: I agree with that. It would be hard to settle on a set of standard app protocols here. Maybe instead it would make sense to have actual validation be limited to a simple character set and max length, but provide guidance on how protocols should be represented. As an example, a potential comment could look like this:

```
The application protocol for this port.
May contain alphanumeric characters as well as `.`, `-`, and `/`.
Must be no longer than 50 characters.
The following application protocols should consistently be represented as:
- HTTP
- HTTPS
- DNS
...
```

I don't know, just throwing something out there.

Comment (howardjohn) [reply pkg/apis/discovery/types.go]:
```
The application protocol for this port.
May contain alphanumeric characters as well as `.`, `-`, and `/`.
Must be no longer than 63 characters.
Examples: HTTP, HTTPS, TCP
```

Would something like that make sense? It still feels a little weird to have a list of "blessed" protocols, even if its just in the docs. I don't care too much either way though

**Review by freehan — APPROVED** [review]
/lgtm
/approve

**Comment by howardjohn** [comment]
Test says `/home/prow/go/src/k8s.io/kubernetes/api/openapi-spec is out of date. Please run hack/update-openapi-spec.sh` but `./build/run.sh hack/update-openapi-spec.sh` makes no changes and I cannot run `hack/update-openapi-spec.sh` directly since I have some version mismatches or something. Not sure what I can do here @robscott

File: cluster/addons/fluentd-elasticsearch/es-image/go.mod
Comment (robscott) [inline cluster/addons/fluentd-elasticsearch/es-image/go.mod]: Not sure how this file got added in (maybe somewhere in the build process?) but I don't think it should be part of this PR.

Comment (howardjohn) [reply cluster/addons/fluentd-elasticsearch/es-image/go.mod]: Thanks! my ide is dumb and keeps adding this

File: pkg/apis/discovery/types.go
Comment (robscott) [reply pkg/apis/discovery/types.go]: I think that strikes a great balance here. There's only so much that can be validated, but this provides clear guidance.

File: pkg/apis/discovery/types.go
Comment (thockin) [inline pkg/apis/discovery/types.go]: nit: linewrap comments at 80

File: pkg/apis/discovery/validation/validation.go
Comment (thockin) [inline pkg/apis/discovery/validation/validation.go]: Confuisingly, you want `IsQualifiedName` - this one tests for values, you need keys.

File: pkg/apis/discovery/validation/validation_test.go
Comment (thockin) [inline pkg/apis/discovery/validation/validation_test.go]: add testcase for `example.com/custom-protocol`

Comment (howardjohn) [reply pkg/apis/discovery/validation/validation_test.go]: Thanks, I added this and addressed your other comments

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Pull request merged**
